src/state_machine.zig

```diff
@@ -2086,6 +1980,136 @@ fn sum_overflows(comptime Int: type, a: Int, b: Int) bool {
     return false;
 }
 
+/// Scans all `Transfers` that already expired at any timestamp.
+/// A custom evaluator is used to stop at the first result where
+/// `expires_at > prefetch_timestamp` while updating the next pulse timestamp.
+/// This way we can achieve the same effect of two conditions with a single scan:
+/// ```
+/// WHERE expires_at <= prefetch_timestamp
+/// UNION
+/// WHERE expires_at > prefetch_timestamp LIMIT 1
+/// ```
+fn ExpirePendingTransfersType(
+    comptime TransfersGroove: type,
+    comptime Storage: type,
+) type {
+    return struct {
+        const ExpirePendingTransfers = @This();
+        const ScanRangeType = @import("lsm/scan_range.zig").ScanRangeType;
+        const EvaluateNext = @import("lsm/scan_range.zig").EvaluateNext;
+        const ScanLookupStatus = @import("lsm/scan_lookup.zig").ScanLookupStatus;
+
+        const Tree = std.meta.FieldType(TransfersGroove.IndexTrees, .expires_at);
+        const Key = Tree.Table.Key;
+        const Value = Tree.Table.Value;
+
+        // TODO(zig) Context should be `*ExpirePendingTransfers`,
+        // but its a dependency loop.
+        const Context = struct {};
+        const Scan = ScanRangeType(
+            Tree,
+            Storage,
+            *Context,
+            value_next,
+            timestamp_from_value,
+        );
+
+        pub const ScanLookup = ScanLookupType(
+            TransfersGroove,
+            Scan,
+            Storage,
+        );
+
+        context: Context = undefined,
+        phase: union(enum) {
+            idle,
+            running: struct {
+                scan: Scan,
+                timestamp: u64,
+            },
+        } = .idle,
+        pulse_timestamp: u64 = TimestampRange.timestamp_min,
```

Comment:
Not sure if these descriptions are correct; please correct me if necessary:

```suggestion
        phase: union(enum) {
            idle,
            running: struct {
                scan: Scan,
                /// Collect all transfers with an expiry less-than-or-equal-to this timestamp.
                timestamp: u64,
            },
        } = .idle,
        /// When ==timestamp_min, there may be pending transfers to expire, but we need to scan to check.
        /// When ==timestamp_max, there are no pending transfers to expire.
        /// Otherwise, this is the timestamp of the next pending transfer expiry.
        pulse_timestamp: u64 = TimestampRange.timestamp_min,
```